SandWorm/Analytics.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Drawing;
+using Rhino.Display;
+using Rhino.Geometry;
+
+namespace SandWorm
+{
+    class Analytics
+    {
+        /// <summary>Implementations of particular analysis options</summary>
+
+        public class None : Analysis.MeshColorAnalysis
+        {
+            public None() : base("No Visualisation") { }
+
+            public override int GetPixelIndexForAnalysis(Point3d vertex, List<Point3d> analysisPts)
+            {
+                return 0; // Should never be called (see below)
+            }
+
+            public override void ComputeLookupTableForAnalysis(double sensorElevation)
+            {
+                lookupTable = new Color[0]; // Empty color table allows pixel loop to skip lookup
+            }
+        }
+
+        public class Elevation : Analysis.MeshColorAnalysis
+        {
+            public Elevation() : base("Visualise Elevation") { }
+
+            public override int GetPixelIndexForAnalysis(Point3d vertex, List<Point3d> analysisPts)
+            {
+                return (int)vertex.Z;
+            }
+
+            public override void ComputeLookupTableForAnalysis(double sensorElevation)
+            {
+                var normalElevationRange = new Analysis.VisualisationRangeWithColor
+                {
+                    ValueStart = 0,
+                    ValueEnd = (int)sensorElevation - 201,
+                    ColorStart = new ColorHSL(0.20, 0.35, 0.02),
+                    ColorEnd = new ColorHSL(0.50, 0.85, 0.85)
+                }; // A clear gradient for pixels inside the expected normal model height 
+
+                var extraElevationRange = new Analysis.VisualisationRangeWithColor
+                {
+                    ValueStart = (int)sensorElevation - 200,
+                    ValueEnd = (int)sensorElevation + 1,
+                    ColorStart = new ColorHSL(1.00, 0.85, 0.76),
+                    ColorEnd = new ColorHSL(0.50, 0.85, 0.99)
+                }; // A fallback gradiend for those outside (TODO: set sensible colors here)
+                ComputeLinearRanges(normalElevationRange, extraElevationRange);
+            }
+        }
+
+        public class Slope : Analysis.MeshColorAnalysis
+        {
+            public Slope() : base("Visualise Slope") { }
+            private int slopeCap = 500; // As measuring in % need an uppper limit on the value
+
+            public override int GetPixelIndexForAnalysis(Point3d vertex, List<Point3d> neighbours)
+            {
+                // Loop over the neighbouring pixels; calculate slopes relative to vertex
+                double slopeSum = 0;
+                for (int i = 0; i < neighbours.Count; i++)
+                {
+                    double rise = vertex.Z - neighbours[i].Z;
+                    double run = Math.Sqrt(Math.Pow(vertex.X - neighbours[i].X, 2) + Math.Pow(vertex.Y - neighbours[i].Y, 2));
```

Comment:
Since we are going with the D4 approach, half of this equation is always going to be zero. In the N-S direction, deltaX = 0, for E-W, deltaY = 0. 
I think this whole line could then be reduced to a simple subtraction. Especially given, that you are checking for the neighbors in the main component anyway. 

```
if (i >= trimmedWidth)
    neighbourPixels.Add(pointCloud[i - trimmedWidth]); // North neighbour
if ((i + 1) % (trimmedWidth) != 0)
    neighbourPixels.Add(pointCloud[i + 1]); // East neighbour
if (i < trimmedWidth * (trimmedHeight - 1))
    neighbourPixels.Add(pointCloud[i + trimmedWidth]); // South neighbour
if (i % trimmedWidth != 0) 
    neighbourPixels.Add(pointCloud[i - 1]); // West neighbour
```